pkg/controller/externalippool/controller.go

```diff
@@ -0,0 +1,417 @@
+	c.consumersWaitGroup.Wait()
+	ipAllocator, exists := c.getIPAllocator(ipPoolName)
+	if !exists {
+		return nil, ErrExternalIPPoolNotFound
+	}
+	ip, err := ipAllocator.AllocateNext()
+	if err == nil {
+		c.queue.Add(ipPoolName)
+	}
+	return ip, err
+}
+
+// AllocateIP allocate an IP from an available IP pool.
+func (c *ExternalIPPoolController) AllocateIP() (net.IP, string, error) {
+	c.consumersWaitGroup.Wait()
+	c.ipAllocatorMutex.RLock()
+	defer c.ipAllocatorMutex.RUnlock()
+	for pool, allocator := range c.ipAllocatorMap {
+		ip, err := allocator.AllocateNext()
+		if err == nil {
+			c.queue.Add(pool)
+			return ip, pool, nil
+		}
+	}
+	return nil, "", errors.New("no ExternalIPPool available")
+}
+
+// UpdateIPAllocation sets the IP in the specified ExternalIPPool.
+func (c *ExternalIPPoolController) UpdateIPAllocation(ip net.IP, pool string) error {
+	ipAllocator, exists := c.getIPAllocator(pool)
+	if !exists {
+		return ErrExternalIPPoolNotFound
+	}
+	err := ipAllocator.AllocateIP(ip)
+	if err == nil {
+		c.queue.Add(pool)
+	}
+	return err
+}
+
+func (c *ExternalIPPoolController) updateExternalIPPoolStatus(poolName string) error {
+	eip, err := c.externalIPPoolLister.Get(poolName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	ipAllocator, exists := c.getIPAllocator(eip.Name)
+	if !exists {
+		return ErrExternalIPPoolNotFound
+	}
+	total, used := ipAllocator.Total(), ipAllocator.Used()
+	toUpdate := eip.DeepCopy()
+	var getErr error
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		actualStatus := eip.Status
+		usage := antreacrds.ExternalIPPoolUsage{Total: total, Used: used}
+		if actualStatus.Usage == usage {
+			return nil
+		}
+		klog.V(2).InfoS("Updating ExternalIPPool status", "ExternalIPPool", poolName, "usage", usage)
+		toUpdate.Status.Usage = usage
+		if _, updateErr := c.crdClient.CrdV1alpha2().ExternalIPPools().UpdateStatus(context.TODO(), toUpdate, metav1.UpdateOptions{}); updateErr != nil && apierrors.IsConflict(updateErr) {
+			toUpdate, getErr = c.crdClient.CrdV1alpha2().ExternalIPPools().Get(context.TODO(), poolName, metav1.GetOptions{})
+			if getErr != nil {
+				return getErr
+			}
+			return updateErr
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("updating ExternalIPPool %s status error: %v", poolName, err)
+	}
+	klog.V(2).InfoS("Updated ExternalIPPool status", "ExternalIPPool", poolName)
+	metrics.AntreaExternalIPPoolStatusUpdates.Inc()
+	return nil
+}
+
+// ReleaseIP releases the IP to the pool.
+func (c *ExternalIPPoolController) ReleaseIP(ip net.IP, poolName string) {
+	allocator, exists := c.getIPAllocator(poolName)
+	if !exists {
+		klog.ErrorS(ErrExternalIPPoolNotFound, "Failed to release IP", "ip", ip, "pool", poolName)
+		return
+	}
+	if err := allocator.Release(ip); err != nil {
+		klog.ErrorS(err, "Failed to release IP", "ip", ip, "pool", poolName)
```

Comment:
Should it let the client handle this error? otherwise the log would look confusing when the method says it fails and the caller says it succeeds.

Reply:
updated.